sqlmesh/core/engine_adapter/clickhouse.py

```diff
@@ -174,6 +175,289 @@ def create_schema(
+            alter_expr.append(
+                "actions",
+                exp.DropPartition(
+                    expressions=[
+                        exp.Partition(
+                            expressions=[
+                                exp.PartitionId(
+                                    this=exp.Literal(
+                                        this=partition, is_string=isinstance(partition, str)
+                                    )
+                                )
+                            ]
+                        )
+                    ],
+                    source=temp_table,
+                ),
+            )
+
+        return alter_expr
+
+    def _replace_by_key(
+        self,
+        target_table: TableName,
+        source_table: QueryOrDF,
+        columns_to_types: t.Optional[t.Dict[str, exp.DataType]],
+        key: t.Sequence[exp.Expression],
+        is_unique_key: bool,
+    ) -> None:
+        source_queries, columns_to_types = self._get_source_queries_and_columns_to_types(
+            source_table, columns_to_types, target_table=target_table
+        )
+
+        key_exp = exp.func("CONCAT_WS", "'__SQLMESH_DELIM__'", *key) if len(key) > 1 else key[0]
+
+        self._insert_overwrite_by_condition(
+            target_table,
+            source_queries,
+            columns_to_types,
+            dynamic_key=key,
+            dynamic_key_exp=key_exp,
+            dynamic_key_unique=is_unique_key,
+        )
+
+    def insert_overwrite_by_partition(
+        self,
+        table_name: TableName,
+        query_or_df: QueryOrDF,
+        partitioned_by: t.List[exp.Expression],
+        columns_to_types: t.Optional[t.Dict[str, exp.DataType]] = None,
+    ) -> None:
+        source_queries, columns_to_types = self._get_source_queries_and_columns_to_types(
+            query_or_df, columns_to_types, target_table=table_name
+        )
+
+        self._insert_overwrite_by_condition(
+            table_name, source_queries, columns_to_types, is_inc_by_partition=True
+        )
+
+    def _create_table_like(
+        self, target_table_name: TableName, source_table_name: TableName
+    ) -> None:
+        """Create table with identical structure as source table"""
+        self.execute(
+            f"CREATE TABLE {target_table_name}{self._on_cluster_sql()} AS {source_table_name}"
+        )
+
+    def _get_affected_partitions(
+        self,
+        table: exp.Table,
+        partition_col_name: str = "_partition_id",
+        where: t.Optional[exp.Condition] = None,
+        limit: t.Optional[int] = None,
+    ) -> t.Set[t.Any]:
+        """List partition IDs present in table"""
+        partitions_query = exp.select(partition_col_name).distinct().from_(table)
+        if where:
+            partitions_query = partitions_query.where(where)
+        if limit:
+            partitions_query = partitions_query.limit(limit)
+        partitions_df = self.fetchdf(partitions_query)
```

Comment:
I noticed that the engine adapter exposes `fetchall()` as well so we can avoid DataFrame's and the associated type mangling if all we are doing is converting resultsets to python sets/lists